DataGrid - AI Column: Implement the DropDownButton Menu

Alyar666 opened the pull request.

Commit:
Add HeaderCellModel

Commit:
small code refactoring

Commit:
Add items for dropdown button

Commit:
Add testcafe test for DropDownButton Menu

Commit:
Add icons. Add click actions

Commit:
Fix copilot remarks

Commit:
Hide DropDownButton when resizing starts

Commit:
minor refactoring of the AI ​​Column architecture

Commit:
Move the addition of the ai command column to the controller

Commit:
Add AIPromptEditorViewController

Commit:
Update DropDownButton items when the prompt changes

Commit:
Fix alignment for DropDownButton Menu

Commit:
add etalons

Commit:
Fix testcafe tests

pharret31 reviewed (dismissed):
My only concern here is about you put the mocks into the component's folder however we had a place to do that - ui/__tests__, did you think about keeping everything in that place? However generally I think the way you did it better

Alyar666 commented:
> My only concern here is about you put the mocks into the component's folder however we had a place to do that - ui/**tests**, did you think about keeping everything in that place? However generally I think the way you did it better

Hi @pharret31,
I added models for components that didn't have their own folder to the ui/tests folder. Otherwise, I added the models to the component folder. I coordinated this with @ksercs.

Raushen commented on packages/devextreme/js/__internal/grids/grid_core/__tests__/__mock__/model/cell/ai_header_cell.ts:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  public getHeaderContent(): HTMLElement | null {
+    return this.root?.querySelector(`.${SELECTORS.aiColumnHeaderContent}`) ?? null;
```
Here is using ai-class, maybe we can use other or rename the model to AIHeaderCellModel? Or make inheritance from base class?

Alyar666 replied:
Yes, you are right. I will rename this model.

Raushen commented on packages/devextreme/js/__internal/grids/grid_core/ai_column/m_ai_column_view.ts:
```diff
@@ -1,197 +1,117 @@
-  ): AIPromptEditorOptions {
-    const alignment = column.alignment === 'right' ? 'left' : 'right';
-    const visibleIndex = this.columnsController.getVisibleIndex(column.index);
+  private activeDropDownButtonInstance!: DropDownButton | null;
```
Can it be null if marked by "!"?

Alyar666 replied:
Yes, it can. `!` - [this modifier](https://scriptdev.ru/guide/028/) tells TypeScript that the variable will definitely be initialized, but later (not in the constructor)

Raushen commented on packages/devextreme/js/__internal/grids/grid_core/ai_column/m_ai_column_view.ts:
```diff
@@ -222,16 +142,52 @@ export const columnHeadersViewExtender = (Base: ModuleType<ColumnHeadersView>) =
+    this.columnsResizer = this.getController('columnsResizer');
+
+    this.columnsResizer.resizeStarted.add(() => {
+      this.activeDropDownButtonInstance?.close();
```
Let's leave a note why this happened

Alyar666 replied:
Ок, I will add it

Commit:
Remove the "dxDataGrid-aiDropDownButtonRegenerateItem" message

vladaskorohodova commented on packages/devextreme/js/localization/messages/en.json:
```diff
@@ -112,6 +112,8 @@
         "dxDataGrid-aiPromptEditorRegenerateButton": "Regenerate Data",
         "dxDataGrid-aiPromptEditorStopButton": "Stop",
+        "dxDataGrid-aiDropDownButtonAutofillItem": "Autofill with AI",
+        "dxDataGrid-aiDropDownButtonClearItem": "Clear Data",
```
```suggestion
        "dxDataGrid-aiDropDownAutofill": "Autofill with AI",
        "dxDataGrid-aiDropDownClear": "Clear Data",
```

Commit:
Rename localization messages

Commit:
Update etalons

Commit:
Add AIHeaderCellModel

Commit:
Add comments for resizeStarted callback
